### DISCUSSION TITLE
A little story about Permit2 amounts

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/v2-periphery#53] [Ideas]
The recent saga with the front-end Permit2 signatures has made an inexplicit-albeit-very-important assumption explicit - and it did it in a big-time way given how much effort @razgraf, @gavriliumircea, and @andreivladbrg have dedicated to debugging the signatures.

The said assumption is that we have been expecting the ERC-20 transfer amount to moonlight as the permitted amount:

- https://github.com/sablierhq/v2-periphery/blob/5e7406e401f9cb90629a183af906c2191b5ff568/src/SablierV2ProxyTarget.sol#L629
- https://github.com/sablierhq/v2-periphery/blob/5e7406e401f9cb90629a183af906c2191b5ff568/src/SablierV2ProxyTarget.sol#L645

This assumption is a foot gun insofar as it doesn't make contract users aware of the role the amount plays in the signature - and this, unfortunately, was the root cause for why the Permit2 signatures failed to work as expected in the front-end for >1 week. See this [unsuspecting line](https://github.com/sablierhq/v2-interfaces/blob/6bdfb4130014a0208bd4be2b14756fccfc0c0026/packages/models/src/Permit/index.ts#L162) in `v2-interfaces`, which hard-codes the permitted amount.

It turns out that the permitted amount need not be equal to the transfer amount. It can be greater than the transfer amount, and @razgraf suggested a potential use case on Slack, i.e.:

> "give me a permit of 10k and at the end of the month I'll trigger a stream of whatever you owe me e.g. 5k"

The dilemma is whether we should bother with this nuanced behavior by adding an `amount` field in the `permit2Params` struct (and potentially add a new check in `_transferAndApprove` that ensures that the permit amount is greater than or equal (`>=`) the transfer amount).

I'm personally undecided. However, I think that it would definitely be helpful to add more explanatory comments in the `_transferAndApprove` function and the `Permit2Params` struct, in order to make the role of `amount` explicit.

_Side note: we are not in a hurry with making this decision - the target is a contract that be easily upgraded (thankfully!), so we could support this behavior in a future release of the periphery system._

**Comment by razgraf** [comment]
I totally agree that, while the exact implementation isn't decide yet (i.e. adding that invariant in code or letting it fail by default due to Permit's own checks) comments around `_transferAndApprove` help a great deal!

**Comment by PaulRBerg** [comment]
Closing as this has been superseded by #54 and #58.
